python/cudf/cudf/core/udf/groupby_typing.py

```diff
@@ -0,0 +1,227 @@
+        self.group_data_type = types.CPointer(group_scalar_type)
+        self.size_type = types.int64
+        self.group_index_type = types.CPointer(index_type)
+        super().__init__(
+            name=f"Group({self.group_scalar_type}, {self.index_type})"
+        )
+
+
+@typeof_impl.register(Group)
+def typeof_group(val, c):
+    """
+    Tie Group and GroupType together such that when Numba
+    sees usage of Group in raw python code, it knows to
+    treat those usages as uses of GroupType
+    """
+    return GroupType(
+        numba.np.numpy_support.from_dtype(val.dtype),
+        numba.np.numpy_support.from_dtype(val.index_dtype),
+    )
+
+
+# The typing of the python "function" Group.__init__
+# as it appears in python code
+@type_callable(Group)
+def type_group(context):
+    def typer(group_data, size, index):
+        if (
+            isinstance(group_data, types.Array)
+            and isinstance(size, types.Integer)
+            and isinstance(index, types.Array)
+        ):
+            return GroupType(group_data.dtype, index.dtype)
+
+    return typer
+
+
+@register_model(GroupType)
+class GroupModel(models.StructModel):
+    def __init__(
+        self, dmm, fe_type
+    ):  # fe_type is fully instantiated group type
+        members = [
+            ("group_data", types.CPointer(fe_type.group_scalar_type)),
+            ("size", types.int64),
+            ("index", types.CPointer(fe_type.index_type)),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+call_cuda_functions: Dict[Any, Any] = {}
+
+
+def _register_cuda_reduction_caller(func, inputty, retty):
+    cuda_func = cuda.declare_device(
+        f"Block{func}_{inputty}", retty(types.CPointer(inputty), types.int64)
+    )
+
+    def caller(data, size):
+        return cuda_func(data, size)
+
+    if call_cuda_functions.get(func.lower()) is None:
+        call_cuda_functions[func.lower()] = {}
+
+    type_key = (retty, inputty)
+    call_cuda_functions[func.lower()][type_key] = caller
+
+
+def _register_cuda_idxreduction_caller(func, inputty):
+    cuda_func = cuda.declare_device(
+        f"Block{func}_{inputty}",
+        types.int64(
+            types.CPointer(inputty), types.CPointer(types.int64), types.int64
+        ),
+    )
+
+    def caller(data, index, size):
+        return cuda_func(data, index, size)
+
+    # only support default index type right now
+    type_key = (index_default_type, inputty)
+    if call_cuda_functions.get(func.lower()) is None:
+        call_cuda_functions[func.lower()] = {}
+    call_cuda_functions[func.lower()][type_key] = caller
+
+
+def _create_reduction_attr(name, retty=None):
+    class Attr(AbstractTemplate):
+        key = name
+
+    if retty:
+
+        def generic(self, args, kws):
+            return nb_signature(retty, recvr=self.this)
+
+    else:
+
+        def generic(self, args, kws):
+            return nb_signature(self.this.group_scalar_type, recvr=self.this)
```

Comment:
Is all this whitespace added by black because it's a function definition? In any case, I'd advocate to simplify the nesting structure:
```suggestion
    if retty is None:
        retty = self.this.group_scalar_type

    def generic(self, args, kws):
        return nb_signature(retty, recvr=self.this)

```

Reply:
Done